Configre keybinds with keynames

TheHussky opened the pull request:
- Add key settings to config
- Use keysyms instead of keycodes
- Bind keysyms to keynames (hardcoded)

NOTE: Map of keynames:keysyms is hardcoded in `keys.hpp` due to limitations of xcb

TheHussky committed:
Add keybinds to config. Use hardcoded keysyms

Add fixes & docs
Add missing symbols
Add a nice typedef to config

mmskv commented on src/config.hpp:
```diff
@@ -15,11 +16,13 @@ using color = const uint32_t;
 const int16_t dxp_x = 0; ///< X coordinate of window's top left corner
 const int16_t dxp_y = 0; ///< Y coordinate of window's top left corner
 
+const bool centering = false;///< Place dxp window centered on screen
+
```
Whats this

TheHussky replied:
leftovers of my centering realization. will delete

mmskv commented on src/config.hpp:
```diff
@@ -15,11 +16,13 @@ using color = const uint32_t;
 ///
 const uint dxp_width = 0;
-const uint dxp_height = 150;
+const uint dxp_height = 250;
```
Restore defaults pls

TheHussky replied:
Why tho? They'll be restored automatically when (if) we merge the branch, won't they?

mmskv replied:
no

mmskv commented on src/config.hpp:
```diff
@@ -50,4 +53,30 @@ const auto dxp_screenshot_period = std::chrono::seconds (1);
+
+const std::vector<int> exit_key = {Escape};
+
+const std::vector<int> choose_key = {Return};
```
Use autoformatter

TheHussky replied:
Done

mmskv commented on src/keynames.hpp:
```diff
@@ -0,0 +1,1389 @@
+const int VoidSymbol=0xFFFFFF; //Does nothing
```
Remove empty comments (or I can do this with couple of key presses in vim)

TheHussky replied:
I think some of them are useful? Is it really needed?
TODO: add all different numbers that are not listed (arabic, farsi, etc.), add latin alphabet (this is for me)

TheHussky replied:
oh, EMPTY comments. Removed

mmskv commented on src/window.cpp:
```diff
@@ -330,15 +329,15 @@ window::handle_event (xcb_generic_event_t *event)
             pres = pres == 0 ? desktops.size () - 1 : pres - 1;
           }
-        if (entr)
+        if (choose)
```
better name would be `select`

TheHussky replied:
Oh, I agree. Will change.

mmskv commented on src/xcb_util.hpp:
```diff
@@ -95,4 +95,8 @@ uint get_current_desktop (xcb_connection_t *c, xcb_window_t root);
 void ewmh_change_desktop (xcb_connection_t *c, xcb_window_t root,
                           uint destkop_id);
 
+xcb_keycode_t* get_keycodes(xcb_connection_t *c, xcb_keysym_t sym);
+
+bool check_key(xcb_connection_t *c, int received_key_code, std::vector<int> key_function);
```
Document functions

TheHussky replied:
Will do

mmskv commented on src/xcb_util.cpp:
```diff
@@ -234,3 +234,42 @@ ewmh_change_desktop (xcb_connection_t *c, xcb_window_t root, uint destkop_id)
+xcb_keycode_t *
+get_keycodes (xcb_connection_t *c, xcb_keysym_t sym) // const char * name)
+{
+  /**
+   * 1. Find a keysym for keyname in struct
+   * 2. Find possible keycodes for keysym
+   * 3. Return a vector of results.
+   */
```
You should place docs above function definition

TheHussky replied:
Will correct

TheHussky replied:
TODO: error handling
P.S. is it considered OK writing things like that in conversations?

mmskv reviewed (changes requested):
- Use autoformatter
- Reset config to defaults

TheHussky committed:
Remove empty comments from keynames.hpp

TheHussky committed:
Remove redundant docs

mmskv commented on src/xcb_util.cpp:
```diff
@@ -234,3 +234,36 @@ ewmh_change_desktop (xcb_connection_t *c, xcb_window_t root, uint destkop_id)
   // of this spec, in which case the timestamp field should be ignored.
   send_xcb_message (c, root, "_NET_CURRENT_DESKTOP", { destkop_id });
 }
+
+xcb_keycode_t *
+get_keycodes (xcb_connection_t *c, xcb_keysym_t sym) // const char * name)
+{
+  auto key_symbols = xcb_key_symbols_alloc (c);
+  if (!key_symbols)
+    {
+      throw std::runtime_error("Couldn't allocate your keyboard symbols");
+    }
+  return xcb_key_symbols_get_keycode (key_symbols, sym);
+}
+
+bool
+check_key (xcb_connection_t *c, int received_key_code,
+           std::vector<int> key_function)
```
Copypaste function docs here too

TheHussky replied:
ok

mmskv commented on src/keynames.hpp:
```diff
@@ -0,0 +1,1509 @@
+const int VoidSymbol = 0xFFFFFF; // Does nothing
```
mb remove all useless keys like VOID_SYMBOL etc?

TheHussky replied:
I'd leave them just to be able to proudly say "Dxp recognizes ALL keys that modmap recognizes"

mmskv commented on src/config.hpp:
```diff
@@ -50,4 +53,30 @@ const auto dxp_screenshot_period = std::chrono::seconds (1);
+/// View keynames.hpp for a list of supported keys
+/// and their names
+
+keys_for action_next = { Down, Right, latin_l, latin_k };
+
+keys_for action_prev = { Left, Up, latin_h, latin_j };
```
Swap j and k buttons. j is next and k is prev

TheHussky replied:
will do

mmskv reviewed (changes requested):
LGTM but too much junk in keynames.hpp

TheHussky committed:
Restore defaults, add docs

mmskv committed:
Restore default config

Rename actions to fit naming scheme

mmskv committed:
Remove redundant comments

mmskv committed:
Merge 'master' into 'keynames'

mmskv committed:
Add constexpr map of keynames : keysyms

mmskv committed:
Parse keyname strings in config

mmskv committed:
Add constexpr map

Fix docs

mmskv committed:
Add constexpr keysyms

Use xcb typedefs instead of builtins.

mmskv committed:
Edit config to be more verbose

mmskv committed:
Add keyname to keycode parser

Most of parsing is constexpr so it is done at compile time.
The only non-constexpr part is getting keycodes from keysyms and it is
ran after window startup.

mmskv committed:
Merge branch 'master' into keynames

mmskv committed:
Cache recent get_hover_desktop. Closes #29

Rename some consts to be more consistent

mmskv committed:
Merge pull request #52 from cyberVirtua/get_hover-optimize

Cache the most recent desktop returned by `get_hover_desktop`

mmskv committed:
Improve codestyle and remove spaghetti

Idk maybe I just obfuscated everything instead :(

mmskv committed:
Run include-what-you-use on the project

mmskv committed:
Sort includes and add missing

Renamed kRequestDesktops to RequestDesktops

mmskv committed:
Merge pull request #53 from cyberVirtua/iwuy

Run Include-what-you-use on the project and apply changes
